include/dxc/Support/dxcapi.use.h

```diff
@@ -31,28 +31,26 @@ class DxcDllSupport {
 
 #ifdef _WIN32
     m_dll = LoadLibraryA(dllName);
-#else
-    m_dll = ::dlopen(dllName, RTLD_LAZY);
-#endif
-
     if (m_dll == nullptr) return HRESULT_FROM_WIN32(GetLastError());
-
-#ifdef _WIN32
     m_createFn = (DxcCreateInstanceProc)GetProcAddress(m_dll, fnName);
-#else
-    m_createFn = (DxcCreateInstanceProc)::dlsym(m_dll, fnName);
-#endif
-
+    
     if (m_createFn == nullptr) {
       HRESULT hr = HRESULT_FROM_WIN32(GetLastError());
-#ifdef _WIN32
       FreeLibrary(m_dll);
+      m_dll = nullptr;
+      return hr;
+    }
 #else
+    m_dll = ::dlopen(dllName, RTLD_LAZY);
+    if (m_dll == nullptr) return E_FAIL;
```

Comment:
Is `E_FAIL` equivalent to `Unspecified error`, as in "we don't know what happened"? And is that what you are referring to with:

> There's a related PR that tries to handle the errors correctly (https://github.com/microsoft/DirectXShaderCompiler/pull/4713), and we should adopt something similar.

... that a _followup_ change should adopt something that more clearly conveys the reason from `dlerror()` or at least make it clear that it is a `dlopen` error and that the caller _could_ check `dlerror()` for more info?

Reply:
Yes. `E_FAIL` is an unspecified error.

Using `dlerror()` would allow us to translate that to a specific error, however `dlerror()` returns errors as strings, and our interface as it is now doesn't really have a good way to propagate the error to the library user (logging to stderr is a pretty terrible way for a library to behave).

I don't think we have clear consensus within the team about how to handle these kinds of errors. My personal opinion is to move away from the Windows-style HRESULT error propagating to a more robust error propagation interface that is cross-platform and more expressive (possibly an extension of `std::error_code` or an `llvm::Error`-like interface).

Right now the best guidance we can give users on Linux is that if library loading fails, they should call `dlerror()` themselves to get the error string.

Reply:
Thanks for making that explicit, it matches what I was assuming.

> Right now the best guidance we can give users on Linux is that if library loading fails, they should call `dlerror()` themselves to get the error string.

Is this documented somewhere, should it be documented somewhere?

Reply:
It is not documented. To my knowledge currently have no documentation for the Linux-specific aspects of our API, so I'm not sure we even have a good place to document it.